hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterStateManager.java

```diff
@@ -113,16 +114,22 @@ void initialClusterState(ClusterState initialState, Version version) {
         clusterServiceLock.lock();
         try {
             node.getNodeExtension().onInitialClusterState(initialState);
+            validateNodeCompatibleWith(version);
 
             final ClusterState currentState = getState();
+            InternalHotRestartService hotRestartService = node.getNodeExtension().getInternalHotRestartService();
+            boolean startingFromPersistence = hotRestartService.isEnabled() && hotRestartService.isStartingFromPersistence();
+            if (startingFromPersistence) {
+                logger.fine("Skipping setting initial cluster state to "
+                        + initialState + ", as instructed by master, because persistence is enabled");
+                return;
```

Comment:
During recovery, the initial cluster state would be set from [`ClusterMetadataManager#prepare`](https://github.com/hazelcast/hazelcast-enterprise/blob/master/hazelcast-enterprise/src/main/java/com/hazelcast/internal/hotrestart/cluster/ClusterMetadataManager.java#L180) and this code was essentially using `currentState != ClusterState.ACTIVE` as a proxy for answering "should cluster state be initialized from master member?". This way, during recovery, the cluster state was set by `ClusterMetadataManager` and was not initialized according to the value received from cluster's master in `FinalizeJoinOp`.

Since with these OS+EE PRs we do not require setting the `ClusterState` as `PASSIVE` cluster-wide during single-member-recovery in k8s, the above check would allow the cluster state received from `FinalizeJoinOp` to be applied, messing with the recovery process. So now the proper check is introduced here, to disallow overwriting the initial cluster state of the rejoining member from master and instead allow it to be managed by `ClusterMetadataManager` until recovery is done.